File: c#/Relativedelta.cs
using System;
 
class Relativedelta {
 
  public int? Year { get; set; } = null;
  public int? Month { get; set; } = null;
  public int? Day { get; set; } = null;
  public int? Hour { get; set; } = null;
  public int? Minute { get; set; } = null;
  public int? Second { get; set; } = null;
 
  private int years = 0;
  private int months = 0;
  private int days = 0;
  private int hours = 0;
  private int minutes = 0;
  private int seconds = 0;
  private int leapdays = 0;
 
  public Relativedelta(DateTime baseDay, DateTime day) {
 
    if (day < baseDay)
      // 日付が基準日より前の場合は例外とする
      throw new ArgumentException();
 
    var months = (day.Year - baseDay.Year) * 12 + (day.Month - baseDay.Month);
    this.SetMonths(months);
 
    var dtm = this.GetDTM(baseDay);
 
    Func<DateTime, DateTime, int> gt = (DateTime day, DateTime dtm) => day.CompareTo(dtm);
    Func<DateTime, DateTime, int> lt = (DateTime day, DateTime dtm) => dtm.CompareTo(day);
    var increment = 0;
    Func<DateTime, DateTime, int>? compare = null;
    if (day < baseDay) {
      compare = gt;
      increment = 1;
    } else {
      compare = lt;
      increment = -1;
    }
 
    while (compare(day, dtm) == 1) {
      months += increment;
      this.SetMonths(months);
      dtm = GetDTM(baseDay);
    }
 
    var delta = day - dtm;
    this.seconds = (int)delta.TotalSeconds;
 
    if (this.seconds > 59) {
      var s = (int)Math.Sign(this.seconds);
      int rem;
      int div = Math.DivRem(this.seconds * s, 60, out rem);
      this.seconds = rem * s;
      this.minutes += div * s;
    }
    if (this.minutes > 59) {
      var s = (int)Math.Sign(this.minutes);
      int rem;
      int div = Math.DivRem(this.minutes * s, 60, out rem);
      this.minutes = rem * s;
      this.hours += div * s;
    }
    if (this.hours > 23) {
      var s = (int)Math.Sign(this.hours);
      int rem;
      int div = Math.DivRem(this.hours * s, 24, out rem);
      this.hours = rem * s;
      this.days += div * s;
    }
    if (months > 11) {
      var s = (int)Math.Sign(this.months);
      int rem;
      int div = Math.DivRem(this.months * s, 12, out rem);
      this.months = rem * s;
      this.years += div * s;
    }
 
    this.Year = this.years;
    this.Month = this.months;
    this.Day = this.days;
    this.Hour = this.hours;
    this.Minute = this.minutes;
    this.Second = this.seconds;
  }
 
  private void SetMonths(int months) {
    this.months = months;
    if (this.months > 11) {
      var s = (int)Math.Sign(this.months);
      int rem;
      int div = Math.DivRem(this.months * s, 12, out rem);
      this.months = rem * s;
      this.years = div * s;
    } else {
      this.years = 0;
    }
  }
 
  private DateTime GetDTM(DateTime other) {
    var y = 0;
    if (this.Year != null) {
      y = (int)this.Year + this.years;
    } else if (other.Year > 0) {
      y = other.Year + this.years;
    }
    var m = this.Month != null ? this.Month : other.Month;
    if (this.months > 0) {
      m += this.months;
      if (m > 12) {
        y += 1;
        m -= 12;
      } else if (m < 1) {
        y -= 1;
        m += 12;
      }
    }
    var _d = this.Day != null ? this.Day : other.Day;
    var d = Math.Min(DateTime.DaysInMonth(y, (int)m), (int)_d);
 
    var repl = (
      year: y,
      month: (int)m,
      day: d,
      hour: (int)(this.Hour != null ? this.Hour : other.Hour),
      minute: (int)(this.Minute != null ? this.Minute : other.Minute),
      second: (int)(this.Second != null ? this.Second : other.Second)
    );
 
    var days = this.days;
    if (this.leapdays > 0 && m > 2 && DateTime.IsLeapYear(y)) {
      days += this.leapdays;
    }
 
    var ret = new DateTime(repl.year, repl.month, repl.day, repl.hour, repl.minute, repl.second);
    return ret.AddDays(days).AddHours(this.hours).AddMinutes(this.minutes).AddSeconds(this.seconds);
  }
}

File: javascript/Relativedelta.js
class Relativedelta {
  constructor(baseDay, day) {
    if (day < baseDay) {
      throw new Error("IllegalArgumentException");
    }
 
    this.year = null;
    this.month = null;
    this.day = null;
    this.hour = null;
    this.minute = null;
    this.second = null;
 
    this.years = 0;
    this.months = 0;
    this.days = 0;
    this.hours = 0;
    this.minutes = 0;
    this.seconds = 0;
    this.leapdays = 0;
 
    let months =
      (day.getFullYear() - baseDay.getFullYear()) * 12 +
      ((day.getMonth() + 1) - (baseDay.getMonth() + 1));
 
    this.setMonths(months);
 
    let dtm = this.getDTM(baseDay);
    // console.log(dtm);
 
    let gt = (a, b) => a > b;
    let lt = (a, b) => a < b;
    let compare = null;
    let increment = 0;
    if (day < baseDay) {
      compare = gt;
      increment = 1;
    } else {
      compare = lt;
      increment = -1;
    }
 
    while (compare(day, dtm)) {
      months += increment;
      this.setMonths(months);
      dtm = this.getDTM(baseDay);
    }
 
    var duration = this.getDuration(day - dtm);
    this.seconds = duration["total_seconds"];
 
    if (this.seconds > 59) {
      var s = Math.sign(this.seconds);
      var div = parseInt((this.seconds * s) / 60);
      var mod = parseInt((this.seconds * s) % 60);
      this.seconds = mod * s;
      this.minutes += div * s;
    }
    if (this.minutes > 59) {
      var s = Math.sign(this.minutes);
      var div = parseInt((this.minutes * s) / 60);
      var mod = parseInt((this.minutes * s) % 60);
      this.minutes = mod * s;
      this.hours += div * s;
    }
    if (this.hours > 23) {
      var s = Math.sign(this.hours);
      var div = parseInt((this.hours * s) / 24);
      var mod = parseInt((this.hours * s) % 24);
      this.hours = mod * s;
      this.days += div * s;
    }
    if (months > 11) {
      var s = Math.sign(this.months);
      var div = parseInt((this.months * s) / 12);
      var mod = parseInt((this.months * s) % 12);
      this.months = mod * s;
      this.years += div * s;
    }
 
    this.year = this.years;
    this.month = this.months;
    this.day = this.days;
    this.hour = this.hours;
    this.minute = this.minutes;
    this.second = this.seconds;
  }
 
  setMonths(months) {
    this.months = months;
    if (this.months > 11) {
      var s = Math.sign(this.months);
      var div = parseInt((this.months * s) / 12);
      var mod = parseInt((this.months * s) % 12);
      this.months = mod * s;
      this.years = div * s;
    } else {
      this.years = 0;
    }
  }
 
  getDTM(other) {
    let y = 0;
    if (this.year != null) {
      y = this.year + this.years;
    } else if (other.getFullYear() > 0) {
      y = other.getFullYear() + this.years;
    }
    let m = this.month != null ? this.month : other.getMonth() + 1;
    if (this.months > 0) {
      m += this.months;
      if (m > 12) {
        y += 1;
        m -= 12;
      } else if (m < 1) {
        y -= 1;
        m += 12;
      }
    }
    let _d = this.day != null ? this.day : other.getDate();
    let d = Math.min(this.lengthOfMonth(y, m), _d);
 
    let repl = {
      "year": y,
      "month": m,
      "day": d
    }
 
    for (const attr of ["hour", "minute", "second"]) {
      if (this[attr] != null) {
        repl[attr] = this[attr];
      }
    }
 
    var days = this.days;
    if (this.leapdays > 0 && m > 2 && isLeap(y)) {
      days += this.leapdays;
    }
 
    let otherObj = {
      "year": other.getFullYear(),
      "month": other.getMonth() + 1,
      "day": other.getDate(),
      "hour": other.getHours(),
      "minute": other.getMinutes(),
      "second": other.getSeconds()
    }
 
    let newObj = { ...otherObj, ...repl };
    let newDate = new Date(
      newObj["year"],
      newObj["month"] - 1,
      newObj["day"] + days,
      newObj["hour"] + this.hours,
      newObj["minute"] + this.minutes,
      newObj["second"] + this.seconds
    );
 
    return newDate;
  }
 
  lengthOfMonth(y, m) {
    let date = new Date(y, m, 0);
    return date.getDate();
  }
 
  isLeap(year) {
    if (year % 4 === 0) {
      if (year % 100 === 0) {
        return year % 400 === 0;
      } else {
        return true;
      }
    }
    return false;
  }
 
  getDuration(duration) {
    let totalMilliseconds = duration;
    let totalSeconds = totalMilliseconds / 1000;
    let totalMinutes = totalSeconds / 60;
    let totalHours = totalMinutes / 60;
    let totalDays = totalHours / 24;
 
    let days = totalDays > 0 ? parseInt(Math.floor(totalDays)) : parseInt(Math.ceil(totalDays))
    let hours = parseInt(totalHours) - (parseInt(totalDays) * 24)
    let minutes = parseInt(parseInt(totalMinutes) - (parseInt(totalHours) * 60))
    let seconds = parseInt(parseInt(totalSeconds) - (parseInt(totalMinutes) * 60))
    let total_seconds = totalSeconds;
 
    return { "days": days, "hours": hours, "minutes": minutes, "seconds": seconds, "total_seconds": total_seconds };
  }
}
 
module.exports = Relativedelta;